Simplify context

ethanfrey opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

I will polish this PR a bit more if it is deemed acceptable, but looking for first responses from devs.

I am getting back into sdk development after some time and find a number of places that seem needless complex - adding unneeded confusion. Going from the go maxim that simpler is better, I would like to offer a number of cleanup PRs. Not sure the best place.

Context is some odd object which is kind of like context.Context but then really like a simple immutible struct with getters and a setter than returns a modified copy of the object. Since the only place to make use of `Value()` and `GetValue()` outside of `context.go` was some governance test code, I figured to just streamline it. I left in an embedded context that can be used like in stdlib for that one case (which maybe is not needed?).

Besides that small change, all cleanup was limited to the `types/context*.go` files. I would like to rename Context to Request, which better represents the meaning of this object, and is in alignment with standard idioms used in go (web) frameworks. I just provide a type alias now to hint at future directions.

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [x] Added a relevant changelog entry: `clog add [section] [stanza] [message]`
- [x] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

jackzampolin commented on .pending/breaking/sdk/4706-Simplify-context:
```diff
@@ -0,0 +1,11 @@
+
+Migration guide:
+`ctx = ctx.WithValue(contextKeyBadProposal, false)` ->
+`ctx = ctx.WithContext(context.WithValue(ctx.Context(), contextKeyBadProposal, false))`
```
Are there any instances of these calls in the code? Seems like based on the diff this is a minimal change.

ethanfrey replied:
in `x/gov/test_common.go` and `x/gov/end_blocker_test.go` to control failure mode or a test handler.

jackzampolin commented on types/context.go:
```diff
@@ -16,167 +13,84 @@ import (
+
+// clone the header before returning
+func (c Context) BlockHeader() abci.Header {
+	// TODO: figure out clone better
```
Guessing this todo will get worked out if you decide to go ahead with this approach.

ethanfrey replied:
yes, it paniced about trying to clone the type, not sure why...

I will ensure there are properly cloned if approved

jackzampolin commented on types/context_test.go:
```diff
@@ -44,18 +44,6 @@ func (l MockLogger) With(kvs ...interface{}) log.Logger {
 	panic("not implemented")
 }
 
-func TestContextGetOpShouldNeverPanic(t *testing.T) {
```
Does this test need to be replicated?

ethanfrey replied:
I deleted the Ops... no longer exist.
Also remove lots of enforced panics

jackzampolin commented on x/gov/endblocker_test.go:
```diff
@@ -309,7 +310,7 @@ func TestEndBlockerProposalHandlerFailed(t *testing.T) {
 	// Create a proposal where the handler will pass for the test proposal
 	// because the value of contextKeyBadProposal is true.
-	ctx = ctx.WithValue(contextKeyBadProposal, true)
+	ctx = ctx.WithContext(context.WithValue(ctx.Context(), contextKeyBadProposal, true))
```
Couldn't we make a simple wrapper for this operation so that we don't break the API here?

ethanfrey replied:
I could...  I wanted to deprecate this usage and enforce a break to see if it is widely used. I can make this non-breaking API wise.

Will update

alessio replied:
I'd love to avoid breaking API here.

jackzampolin commented:
This gets a 👍 on direction from me. Seems to make the code much cleaner! Thanks for the contribution!

rigelrozanski commented on types/context.go:
```diff
@@ -16,167 +13,84 @@ import (
-		return proto.Clone(message)
-	}
-	return value
+	ctx           context.Context
```
I don't see the advantage of having `Context` no longer extend `context.Context` - I would prefer to see this change reversed

alessio replied:
I agree with @rigelrozanski here

ethanfrey replied:
I don't see any advantages there. I will discuss usage of Context elsewhere, but technically

```go
type Context interface {
	Deadline() (deadline time.Time, ok bool)
	Done() <-chan struct{}
	Err() error
	Value(key interface{}) interface{}
}
```

Only `Value` is explicitly implemented by `sdk.Context`. If we try to use it as a normal Context object, with any of the standard functions, we quickly get issues. (This code was checked against the master branch):

```go
func demo() {
	var ctx Context
	ctx, _ = context.WithCancel(ctx)
}
```

-> `types/context.go:64:9: cannot assign context.Context to ctx (type Context) in multiple assignment: need type assertion`

So we pretend to be a `context.Context`, but if we pass it in like that, we are no longer `sdk.Context`. Which means embedding it this was is actually *dangerous*.

By including Context as a member, we can then properly do the above desired issue (this code off of the PR):

```go
func demo() {
	var ctx Context
	std, cancel := context.WithTimeout(ctx.Context(), time.Second)
	defer cancel()
	ctx = ctx.WithContext(std)
	select {
	case <-ctx.Context().Done():
			fmt.Println("Done!")
	}
}
```

Note that the compiler does not let me try to pass in `sdk.Context` as `context.Context`, which is a benefit.

ethanfrey replied:
I also think sdk.Context is a misnomer. 
It is really more like a Request.

I prefer Context for holding non-necessary values, not critical points like the DB itself. (not to say other frameworks, like gin, don't use this naming of Context as well)

alessio replied:
I think I've changed my mind.

> So we pretend to be a context.Context, but if we pass it in like that, we are no longer sdk.Context. Which means embedding it this was is actually dangerous.

I always thought that Golang's context was a pretty clunky solution to two problems:
1. HTTP request-specific values
2. cancellation

IMvHO Golang does not elegantly solve either. Plus, our custom Context was not even designed to address 2. and Golang 2 will supposedly solve 2. once and for all.

In conclusion:

* ACK on `Context -> Request` renaming
* ACK on the removal of the composition with `context.Context`

ethanfrey replied:
> I always thought that Golang's context was a pretty clunky solution to two problems:
>   1. HTTP request-specific values
>   2. cancellation

I agree with you here. It does work as a generic store (1), but that is usually a sign of a design flaw. I mean a `map` or `array` with a specific type could often be a better solution (as you know in advance which way you want to extend). That said, I don't think we need to remove that functionality. I would rather deprecate it until we ensure there are no needs for it.

The only real bonus I find is that some frameworks like [monkit](https://github.com/spacemonkeygo/monkit) will "auto"-instrument your code if you pass them Context. I see main value just for interoperating with other libs.

ethanfrey replied:
@rigelrozanski if the issue is verbose-ness (which is true), I am happy to rename the methods to `Ctx()` and `WithCtx()`, which make it shorter, but possibly less clear read.

rigelrozanski replied:
Sorry I couldn't follow up on this thread took some vacation! Sounds fine though

alessio commented on .pending/breaking/sdk/4706-Simplify-context:
```diff
@@ -0,0 +1,11 @@
+Only breaking change is not to support `Value` and `GetValue` as first class calls.
+We do embed ctx.Context() as a raw context.Context instead to be used as you see fit.
+
+Migration guide:
```
I'm deep in love these few lines, thanks @ethanfrey!

ethanfrey commented:
I addressed all of the code comments here, except for the protobuf Clone.
I will discuss the design comment above in an issue.

ethanfrey commented:
Thanks @alessio I will do a deeper read through docs and polish of the last edges here.

Glad to see you like the rename as well. I just added an optional type alias here:
`type Request = Context`
so people can refer to it as sdk.Request in their Handlers if they wish.

But if there is support, I would make a separate PR for a larger rename, updating all references to Context in sdk, but keep backwards compatibility with external libs with:
`type Context = Request`

ethanfrey commented:
I fixed the protobuf clone issue, seems to be a time.Time problem.
And shows they were never actually Cloned in the older code

ethanfrey commented:
`Context` appears many places in the docs (so renaming to Request is a big work).
I did minor godoc updates.

Otherwise, I think this is done

ethanfrey commented:
Any progress on the review process? I think I addressed all concerns raised.

I also believe it is no longer API breaking, as I did add `Value()` and `WithValue()` back to `sdk.Context`. It doesn't inherit the other fields from Context (such as Done() and Error()), but I highly doubt anyone was using those, so I am unsure of the API breaking nature of this PR.

fedekunze commented:
@ethanfrey, someone with merge rights will have to merge it cc: @alessio @jackzampolin @alexanderbez @rigelrozanski. I believe Bez will want to review this once he comes back from vacations tomorrow

ethanfrey commented:
Thanks @alessio

alexanderbez commented:
Took a look at the PR and the changes look reasonable to me, although I'm not sure I'm necessarily in favor of having to require the `sdk.Context` contain fields for the primary getters (e.g. txBytes). I suppose this is fine for now.

Also, what is the harm of using `ctx.WithValue(...)` over the new preferred API where I have to manually get the context? In the SDK, I cannot see where the API would be preferred.

ethanfrey commented:
Hi @alexanderbez my point was that we only use this With Value in some governance test code, and in general, we should not encourage this usage of arbitrary data, when we really use it for structured data.

I left the ability to use it as such and no breaking api changes, but push towards a clear and simple struct which is easy to understand. In general, the simplest solution that satisfies the requirements is easier to understand, easier to change later and brings less technical debt.

This seemed like low hanging fruit to clean up, as I was not sure all the ways it was used in the Sdk without searching a lot
